python-lib/tagger.py

```diff
@@ -1,44 +1,60 @@
-"""Module to prepare the tagging of the documents
-splits on sentences and instanciates a Matcher to tag the documents"""
+"""Main module to tag the documents"""
 
 from spacy_tokenizer import MultilingualTokenizer
 from dkulib_io_utils import generate_unique
 from spacy.matcher import PhraseMatcher
 from fastcore.utils import store_attr
-from typing import AnyStr
+from typing import AnyStr, List
 import pandas as pd
+from tagger_formatter import (
+    FormatterByDocumentJson,
+    FormatterByDocument,
+    FormatterByTag,
+)
+from enum import Enum
+from time import perf_counter
+import logging
+
+
+class OutputFormat(Enum):
+    ONE_ROW_PER_TAG = "one_row_per_tag"
+    ONE_ROW_PER_DOCUMENT = "one_row_per_doc"
+    ONE_ROW_PER_DOCUMENT_JSON = "one_row_per_doc_json"
 
 
 class Tagger:
     def __init__(
         self,
-        text_df,
-        ontology_df,
-        text_column,
-        language,
-        language_column,
-        tag_column,
-        category_column,
-        keyword_column,
-        lemmatize,
-        case_sensitive,
-        normalize,
-        mode,
+        text_df: pd.DataFrame,
+        ontology_df: pd.DataFrame,
+        text_column: AnyStr,
+        language: AnyStr,
+        language_column: AnyStr,
+        tag_column: AnyStr,
+        category_column: AnyStr,
+        keyword_column: AnyStr,
+        lemmatization: bool,
+        case_insensitive: bool,
+        normalization: bool,
+        mode: AnyStr,
     ):
-        store_attr(but=["text_df", "ontology_df"])
-        self.text_df = text_df.get_dataframe()
-        self.ontology_df = ontology_df.get_dataframe()
+        store_attr()
         self.matcher_dict = {}
         self.nlp_dict = {}
-        self._matching_pipeline()
 
-    def _get_patterns(self):
+    def _get_patterns(self) -> None:
         """
         Public function called in tagger.py
         Creates the list of patterns :
         - If there aren't category -> list of the keywords (string list)
         - If there are categories  -> list of dictionaries, {"label": category, "pattern": keyword}
         """
+        # remove rows with missing values
+        self.ontology_df.replace("", float("nan"), inplace=True)
+        self.ontology_df.dropna(inplace=True)
+        assert not (
```

Comment:
Nitpick: `assert` is usually used for developer-oriented code like unit tests. Since this error is meant for an end-user, the right expectation is `ValueError`.

Also, there is no concept like "restarting a plugin". I guess you mean "re-run the recipe"?

Reply:
that's what I meant indeed :) here fb5c4d0